test/test_fork.py

```diff
@@ -0,0 +1,147 @@
+class TestFork(IntegrationTest):
+    def setUp(self):
+        self.db = self.client.pymongo_test
+        self.fork_thread = ForkThread()
+
+    def test_lock_client(self):
+        """
+        Forks the client with some items locked.
+        Parent => All locks should be as before the fork.
+        Child => All locks should be reset.
+        """
+
+        def exit_cond():
+            # Checking all the locks after forking this way is highly
+            # probabilistic as more locking behavior may ensue. Instead we
+            # check that we can acquire _insertion_lock, meaning we have
+            # successfully completed.
+            with _ForkLock._insertion_lock:
+                return 0  # success
+
+        self.fork_thread.exit_cond = exit_cond
+        with patch.object(
+            self.db.client, "_MongoClient__lock", LockWrapper(fork_thread=self.fork_thread)
+        ):
+            # Call _get_topology, will launch a thread to fork upon __enter__ing
+            # the with region.
+            self.db.client._get_topology()
+            self.db.client._MongoClient__lock.fork_thread.join()
+            lock_pid: int = self.db.client._MongoClient__lock.fork_thread.pid
+            # The POSIX standard states only the forking thread is cloned.
+            # In the parent, it'll return here.
+            # In the child, it'll end with the calling thread.
+            self.assertNotEqual(lock_pid, 0)
+            self.assertEqual(0, os.waitpid(lock_pid, 0)[1] >> 8)
+
+    def test_lock_object_id(self):
+        """
+        Forks the client with ObjectId's _inc_lock locked.
+        Will fork upon __enter__, waits for child to return.
+        Parent => _inc_lock should remain locked.
+        Child => _inc_lock should be unlocked.
+        Must use threading.Lock as ObjectId uses this.
+        """
+        self.fork_thread.exit_cond = lambda: 0 if not ObjectId._inc_lock.locked() else 1
+        with patch.object(
+            ObjectId,
+            "_inc_lock",
+            LockWrapper(fork_thread=self.fork_thread, lock_type=threading.Lock),
+        ):
+            # Generate the ObjectId, will generate a thread to fork
+            # upon __enter__ing the with region.
+
+            ObjectId()
+            ObjectId._inc_lock.fork_thread.join()
+            lock_pid: int = ObjectId._inc_lock.fork_thread.pid
+
+            self.assertNotEqual(lock_pid, 0)
+            self.assertEqual(0, os.waitpid(lock_pid, 0)[1] >> 8)
+
+    def test_topology_reset(self):
+        """
+        Tests that topologies are different from each other.
+        Cannot use ID because virtual memory addresses may be the same.
+        Cannot reinstantiate ObjectId in the topology settings.
+        Relies on difference in PID when opened again.
+        """
+        parent_conn, child_conn = Pipe()
+        cl_test = MongoClient()
+        init_id = cl_test._topology._pid
+        lock_pid: int = os.fork()
+
+        if lock_pid == 0:  # Child
```

Comment:
For the Topology, I think the `_pid` will always change, and the `_kill_cursors_executor._thread.ident` will change as well.

Reply:
I don't think the `thread.ident` one will work. Thread IDs wrt processes can be recycled, and the only way to get a thread's ID wrt the kernel is to be calling `threading.get_native_id()` on a desired thread on Python 3.8. 

Is there another way it might be possible to verify `_kill_cursors_executor` has been reinitialized? I don't see anything in `PeriodicExecutor` whose state would be distinct upon restarting in a child process

Reply:
What about:
```
parent_cursor_exc = client._kill_cursors_executor
...
child_conn.send((parent_cursor_exc != client._kill_cursors_executor, "client._kill_cursors_executor was not reinitialized"))

...

passed, msg = parent_conn.recv()
self.assertTrue(passed, msg)
```